File: 0502-ipo/0502-ipo.java
class Solution {
    class Pair implements Comparable<Pair> {
        int capital, profit;
        
        public Pair(int capital, int profit) {
            this.capital = capital;
            this.profit = profit;
        }

        public int compareTo(Pair pair) {
            return capital - pair.capital;
        }
    }

    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        Pair[] projects = new Pair[n];
        for(int i = 0; i<n; i++) {
            projects[i] = new Pair(capital[i], profits[i]);
        }
        Arrays.sort(projects);

        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        int idx = 0;

        for(int i = 0; i<k; i++) {
            while(idx < n && projects[idx].capital <= w) {
                pq.add(projects[idx].profit);
                idx++;
            }

            if(!pq.isEmpty()) w += pq.poll();
        }

        return w;
    }
}

File: 0502-ipo/0502-ipo.py
class Solution(object):
    def my_compare(a, b):
        if a[0] == b[0]:
            return a[1] > b[1]
        return a[0] < b[0]
        
    def findMaximizedCapital(self, k, w, profits, capital):
        n = len(profits)
        projects = [(capital[i], profits[i]) for i in range(n)]
        projects.sort(key=lambda x: (x[0], -x[1]))

        idx = 0
        max_heap = []
        while k > 0:
            while idx < n and projects[idx][0] <= w:
                heapq.heappush(max_heap, -projects[idx][1])
                idx += 1
            
            if max_heap:
                w += -heapq.heappop(max_heap)
            k -= 1

        return w
        